// rolldown/dist/cli.mjs
e -f iife -e jQuery,window._ -g jQuery=$"
	}
];
const notes = ["CLI options will override the configuration file.", "For more information, please visit https://rolldown.rs/."];
/**
* Generates the CLI help text as a string.
*/
function generateHelpText() {
	const lines = [];
	lines.push(`${styleText("gray", `${description} (rolldown v${version})`)}`);
	lines.push("");
	lines.push(`${styleText(["bold", "underline"], "USAGE")} ${styleText("cyan", "rolldown -c <config>")} or ${styleText("cyan", "rolldown <input> <options>")}`);
	lines.push("");
	lines.push(`${styleText(["bold", "underline"], "OPTIONS")}`);
	lines.push("");
	lines.push(Object.entries(options).sort(([a], [b]) => {
		if (options[a].short && !options[b].short) return -1;
		if (!options[a].short && options[b].short) return 1;
		if (options[a].short && options[b].short) return options[a].short.localeCompare(options[b].short);
		return a.localeCompare(b);
	}).map(([option, { type, short, hint, description }]) => {
		let optionStr = `  --${option} `;
		if (short) optionStr += `-${short}, `;
		if (type === "string") optionStr += `<${hint ?? option}>`;
		if (description && description.length > 0) description = description[0].toUpperCase() + description.slice(1);
		return styleText("cyan", optionStr.padEnd(30)) + description + (description && description?.endsWith(".") ? "" : ".");
	}).join("\n"));
	lines.push("");
	lines.push(`${styleText(["bold", "underline"], "EXAMPLES")}`);
	lines.push("");
	examples.forEach(({ title, command }, ord) => {
		lines.push(`  ${ord + 1}. ${title}:`);
		lines.push(`    ${styleText("cyan", command)}`);
		lines.push("");
	});
	lines.push(`${styleText(["bold", "underline"], "NOTES")}`);
	lines.push("");
	notes.forEach((note) => {
		lines.push(`  * ${styleText("gray", note)}`);
	});
	return lines.join("\n");
}
function showHelp() {
	logger.log(generateHelpText());
}
//#endregion
//#region src/cli/version-check.ts
function checkNodeVersion(nodeVersion) {
	const currentVersion = nodeVersion.split(".");
	const major = parseInt(currentVersion[0], 10);
	const minor = parseInt(currentVersion[1], 10);
	return major === 20 && minor >= 19 || major === 22 && minor >= 12 || major > 22;
}
//#endregion
//#region src/cli/index.ts
if (!checkNodeVersion(g$1.versions.node)) logger.warn(`You are using Node.js ${g$1.versions.node}. Rolldown requires Node.js version 20.19+ or 22.12+. Please upgrade your Node.js version.`);
async function main() {
	const { rawArgs, ...cliOptions } = parseCliArguments();
	if (cliOptions.environment) {
		const environment = Array.isArray(cliOptions.environment) ? cliOptions.environment : [cliOptions.environment];
		for (const argument of environment) for (const pair of argument.split(",")) {
			const [key, ...value] = pair.split(":");
			g$1.env[key] = value.length === 0 ? String(true) : value.join(":");
		}
	}
	if (cliOptions.help) {
		showHelp();
		return;
	}
	if (cliOptions.version) {
		logger.log(`rolldown v${version}`);
		return;
	}
	if (cliOptions.config || cliOptions.config === "") {
		await bundleWithConfig(cliOptions.config, cliOptions, rawArgs);
		return;
	}
	if ("input" in cliOptions.input) {
		await bundleWithCliOptions(cliOptions);
		return;
	}
	showHelp();
}
main().catch((err) => {
	logger.error(err);
	g$1.exit(1);
});
//#endregion
export {};
esult = {
		type: info.type === "boolean" ? "boolean" : "string",
		description
	};
	if (config?.abbreviation) result.short = config.abbreviation;
	if (config?.hint) result.hint = config.hint;
	return [config?.reverse ? `no-${key}` : key, result];
}));
const knownKeys = new Set(Object.keys(schemaInfo));
for (const key of Object.keys(schemaInfo)) {
	const dotIdx = key.indexOf(".");
	if (dotIdx > 0) knownKeys.add(key.substring(0, dotIdx));
}
const shortAliases = /* @__PURE__ */ new Set();
for (const config of Object.values(alias)) if (config?.abbreviation) shortAliases.add(config.abbreviation);
function kebabToCamelCase(input) {
	return input.replaceAll(/([a-z])-([a-z])/g, (_, p1, p2) => {
		return p1 + p2.toUpperCase();
	});
}
function camelizeNestedKeys(value) {
	const result = {};
	for (const [key, nestedValue] of Object.entries(value)) if (Array.isArray(nestedValue)) result[kebabToCamelCase(key)] = nestedValue;
	else if (nestedValue && typeof nestedValue === "object") result[kebabToCamelCase(key)] = camelizeNestedKeys(nestedValue);
	else result[kebabToCamelCase(key)] = nestedValue;
	return result;
}
function parseCliArguments() {
	const cli = cac("rolldown");
	for (const [key, info] of Object.entries(schemaInfo)) {
		if (info.type === "never") continue;
		const config = alias[key];
		let rawName = "";
		if (config?.abbreviation) rawName += `-${config.abbreviation}, `;
		if (config?.reverse) rawName += `--no-${key}`;
		else rawName += `--${key}`;
		if (info.type !== "boolean" && !config?.reverse) if (config?.requireValue) rawName += ` <${config?.hint ?? key}>`;
		else rawName += ` [${config?.hint ?? key}]`;
		cli.option(rawName, info.description ?? config?.description ?? "");
	}
	let parsedInput = [];
	let parsedOptions = {};
	const cmd = cli.command("[...input]", "");
	cmd.allowUnknownOptions();
	cmd.ignoreOptionDefaultValue();
	cmd.action((input, opts) => {
		parsedInput = input;
		parsedOptions = opts;
	});
	try {
		cli.parse(process.argv, { run: true });
	} catch (err) {
		if (err?.name === "CACError") {
			const match = err.message.match(/option `(.+?)` value is missing/);
			if (match) {
				const optName = match[1].replace(/ [<[].*/, "").replace(/^-\w, /, "");
				logger.error(`Option \`${optName}\` requires a value but none was provided.`);
			} else logger.error(err.message);
			process.exit(1);
		}
		throw err;
	}
	delete parsedOptions["--"];
	for (const short of shortAliases) delete parsedOptions[short];
	for (const key of Object.keys(parsedOptions)) if (key === "__proto__" || key === "constructor" || key === "prototype" || key.startsWith("__proto__.") || key.startsWith("constructor.") || key.startsWith("prototype.")) delete parsedOptions[key];
	const unknownKeys = Object.keys(parsedOptions).filter((k) => !knownKeys.has(k));
	if (unknownKeys.length > 0) {
		unknownKeys.sort();
		const single = unknownKeys.length === 1;
		logger.warn(`Option \`${unknownKeys.join(",")}\` ${single ? "is" : "are"} unrecognized. We will ignore ${single ? "this" : "those"} option${single ? "" : "s"}.`);
	}
	parsedOptions = camelizeNestedKeys(parsedOptions);
	const rawArgs = { ...parsedOptions };
	for (const key of unknownKeys) delete parsedOptions[key];
	for (const [key, value] of Object.entries(parsedOptions)) {
		const type = schemaInfo[key]?.type;
		if (Array.isArray(value)) {
			if (type !== "array" && type !== "object") parsedOptions[key] = value[value.length - 1];
		} else if (type === "array" && typeof value === "string") parsedOptions[key] = [value];
	}
	for (const [schemaKey, info] of Object.entries(schemaInfo)) {
		if (info.type !== "object") continue;
		const parts = schemaKey.split(".");
		let parent = parsedOptions;
		for (let i = 0; i < parts.length - 1; i++) parent = parent?.[parts[i]];
		const leafKey = parts[parts.length - 1];
		const value = parent?.[leafKey];
		if (value === void 0) continue;
		const values = Array.isArray(value) ? value : [value];
		if (typeof values[0] !== "string") continue;
		let usedDeprecatedSyntax = false;
		const result = {};
		for (const v of values) for (const pair of String(v).split(",")) {
			const colonIdx = pair.indexOf(":");
			const eqIdx = pair.indexOf("=");
			let k;
			let val;
			if (colonIdx > 0 && (eqIdx === -1 || colonIdx < eqIdx)) {
				k = pair.slice(0, colonIdx);
				val = pair.slice(colonIdx + 1);
			} else if (eqIdx > 0) {
				k = pair.slice(0, eqIdx);
				val = pair.slice(eqIdx + 1);
				usedDeprecatedSyntax = true;
			} else continue;
			result[k] = val;
		}
		if (usedDeprecatedSyntax) {
			const optionName = camelCaseToKebabCase(schemaKey);
			logger.warn(`Using \`key=value\` syntax for \`--${optionName}\` is deprecated. Use \`key:value\` instead.`);
		}
		parent[leafKey] = result;
	}
	return {
		...normalizeCliOptions(parsedOptions, parsedInput),
		rawArgs
	};
}
//#endregion
//#region src/utils/clear-screen.ts
const CLEAR_SCREEN = "\x1Bc";
function getClearScreenFunction(options) {
	const isTTY = process.stdout.isTTY;
	const isAnyOptionNotAllowingClearScreen = arraify(options).some(({ watch }) => watch === false || watch?.clearScreen === false);
	if (isTTY && !isAnyOptionNotAllowingClearScreen) return () => {
		process.stdout.write(CLEAR_SCREEN);
	};
}
//#endregion
//#region \0@oxc-project+runtime@0.139.0/helpers/esm/usingCtx.js
function _usingCtx() {
	var r = "function" == typeof SuppressedError ? SuppressedError : function(r, e) {
		var n = Error();
		return n.name = "SuppressedError", n.error = r, n.suppressed = e, n;
	}, e = {}, n = [];
	function using(r, e) {
		if (null != e) {
			if (Object(e) !== e) throw new TypeError("using declarations can only be used with objects, functions, null, or undefined.");
			if (r) var o = e[Symbol.asyncDispose || Symbol["for"]("Symbol.asyncDispose")];
			if (void 0 === o && (o = e[Symbol.dispose || Symbol["for"]("Symbol.dispose")], r)) var t = o;
			if ("function" != typeof o) throw new TypeError("Object is not disposable.");
			t && (o = function o() {
				try {
					t.call(e);
				} catch (r) {
					return Promise.reject(r);
				}
			}), n.push({
				v: e,
				d: o,
				a: r
			});
		} else r && n.push({
			d: e,
			a: r
		});
		return e;
	}
	return {
		e,
		u: using.bind(null, !1),
		a: using.bind(null, !0),
		d: function d() {
			var o, t = this.e, s = 0;
			function next() {
				for (; o = n.pop();) try {
					if (!o.a && 1 === s) return s = 0, n.push(o), Promise.resolve().then(next);
					if (o.d) {
						var r = o.d.call(o.v);
						if (o.a) return s |= 2, Promise.resolve(r).then(next, err);
					} else s |= 1;
				} catch (r) {
					return err(r);
				}
				if (1 === s) return t !== e ? Promise.reject(t) : Promise.resolve();
				if (t !== e) throw t;
			}
			function err(n) {
				return t = t !== e ? new r(n, t) : n, next();
			}
			return next();
		}
	};
}
//#endregion
//#region src/cli/commands/bundle.ts
async function bundleWithConfig(configPath, cliOptions, rawArgs = {}) {
	if (cliOptions.watch) {
		process.env.ROLLUP_WATCH = "true";
		process.env.ROLLDOWN_WATCH = "true";
	}
	const config = await loadConfig(configPath, { configLoader: cliOptions.configLoader });
	const resolvedConfig = typeof config === "function" ? await config(rawArgs) : config;
	if (typeof resolvedConfig !== "object" || resolvedConfig === null) {
		logger.error(`Invalid configuration from ${configPath}: expected object or array, got ${resolvedConfig}`);
		process.exit(1);
	}
	if (cliOptions.watch) await watchInner(resolvedConfig, cliOptions);
	else await bundleInner(resolvedConfig, cliOptions);
}
async function bundleWithCliOptions(cliOptions) {
	try {
		var _usingCtx$1 = _usingCtx();
		if (cliOptions.output.dir || cliOptions.output.file) {
			await (cliOptions.watch ? watchInner : bundleInner)({}, cliOptions);
			return;
		}
		if (cliOptions.watch) {
			logger.error("You must specify `output.dir` to use watch mode");
			process.exit(1);
		}
		const { output: outputs } = await _usingCtx$1.a(await rolldown(cliOptions.input)).generate(cliOptions.output);
		if (outputs.length === 0) {
			logger.error("No output generated");
			process.exit(1);
		}
		for (const file of outputs) {
			if (outputs.length > 1) logger.log(`\n${styleText(["cyan", "bold"], `|→ ${file.fileName}:`)}\n`);
			console.log(file.type === "asset" ? file.source : file.code);
		}
	} catch (_) {
		_usingCtx$1.e = _;
	} finally {
		await _usingCtx$1.d();
	}
}
async function watchInner(config, cliOptions) {
	let normalizedConfig = arraify(config).map((option) => {
		return {
			...option,
			...cliOptions.input,
			output: arraify(option.output || {}).map((output) => {
				return {
					...output,
					...cliOptions.output
				};
			})
		};
	});
	const watcher = watch(normalizedConfig);
	onExit((code) => {
		Promise.resolve(watcher.close()).finally(() => {
			process.exit(typeof code === "number" ? code : 0);
		});
		return true;
	});
	const changedFile = [];
	watcher.on("change", (id, event) => {
		if (event.event === "update") changedFile.push(id);
	});
	const clearScreen = getClearScreenFunction(normalizedConfig);
	watcher.on("event", async (event) => {
		switch (event.code) {
			case "START":
				clearScreen?.();
				break;
			case "BUNDLE_START":
				if (changedFile.length > 0) logger.log(`Found ${styleText("bold", changedFile.map(relativeId).join(", "))} changed, rebuilding...`);
				changedFile.length = 0;
				break;
			case "BUNDLE_END":
				await event.result.close();
				logger.success(`Rebuilt ${styleText("bold", relativeId(event.output[0]))} in ${styleText("green", ms(event.duration))}.`);
				break;
			case "ERROR":
				await event.result.close();
				logger.error(event.error);
				break;
			default: break;
		}
	});
	logger.log(`Waiting for changes...`);
}
async function bundleInner(config, cliOptions) {
	const startTime = performance.now();
	const result = [];
	const configList = arraify(config);
	for (const config of configList) {
		const outputList = arraify(config.output || {});
		const build = await rolldown({
			...config,
			...cliOptions.input
		});
		try {
			for (const output of outputList) result.push(await build.write({
				...output,
				...cliOptions.output
			}));
		} finally {
			await build.close();
		}
	}
	result.forEach(printBundleOutputPretty);
	logger.log(``);
	const duration = performance.now() - startTime;
	logger.success(`rolldown v${version} Finished in ${styleText("green", ms(duration))}`);
}
function printBundleOutputPretty(output) {
	const outputEntries = collectOutputEntries(output.output);
	printOutputEntries(outputEntries, collectOutputLayoutAdjustmentSizes(outputEntries), "<DIR>");
}
function collectOutputEntries(output) {
	return output.map((chunk) => ({
		type: chunk.type,
		fileName: chunk.fileName,
		size: chunk.type === "chunk" ? Buffer.byteLength(chunk.code) : Buffer.byteLength(chunk.source)
	}));
}
function collectOutputLayoutAdjustmentSizes(entries) {
	let longest = 0;
	let biggestSize = 0;
	for (const entry of entries) {
		if (entry.fileName.length > longest) longest = entry.fileName.length;
		if (entry.size > biggestSize) biggestSize = entry.size;
	}
	const sizePad = displaySize(biggestSize).length;
	return {
		longest,
		biggestSize,
		sizePad
	};
}
const numberFormatter = new Intl.NumberFormat("en", {
	maximumFractionDigits: 2,
	minimumFractionDigits: 2
});
function displaySize(bytes) {
	return `${numberFormatter.format(bytes / 1e3)} kB`;
}
const CHUNK_GROUPS = [{
	type: "asset",
	color: "green"
}, {
	type: "chunk",
	color: "cyan"
}];
function printOutputEntries(entries, sizeAdjustment, distPath) {
	for (const group of CHUNK_GROUPS) {
		const filtered = entries.filter((e) => e.type === group.type);
		if (!filtered.length) continue;
		for (const entry of filtered.sort((a, z) => a.size - z.size)) {
			let log = styleText("dim", withTrailingSlash(distPath));
			log += styleText(group.color, entry.fileName.padEnd(sizeAdjustment.longest + 2));
			log += styleText("dim", entry.type);
			log += styleText("dim", ` │ size: ${displaySize(entry.size).padStart(sizeAdjustment.sizePad)}`);
			logger.log(log);
		}
	}
}
function withTrailingSlash(path) {
	if (path[path.length - 1] !== "/") return `${path}/`;
	return path;
}
function ms(duration) {
	return duration < 1e3 ? `${duration.toFixed(2)} ms` : `${(duration / 1e3).toFixed(2)} s`;
}
function relativeId(id) {
	if (!path.isAbsolute(id)) return id;
	return path.relative(path.resolve(), id);
}
//#endregion
//#region src/cli/commands/help.ts
const examples = [
	{
		title: "Bundle with a config file `rolldown.config.mjs`",
		command: "rolldown -c rolldown.config.mjs"
	},
	{
		title: "Bundle the `src/main.ts` to `dist` with `cjs` format",
		command: "rolldown src/main.ts -d dist -f cjs"
	},
	{
		title: "Bundle the `src/main.ts` and handle the `.png` assets to Data URL",
		command: "rolldown src/main.ts -d dist --moduleTypes .png=dataurl"
	},
	{
		title: "Bundle the `src/main.tsx` and minify the output with sourcemap",
		command: "rolldown src/main.tsx -d dist -m -s"
	},
	{
		title: "Create self-executing IIFE using external jQuery as `$` and `_`",
		command: "rolldown src/main.ts -d dist -n bundl
