// rolldown/dist/utils-index.mjs
", 149],
	["TSSatisfiesExpression", 150],
	["TSTemplateLiteralType", 151],
	["TSTupleType", 152],
	["TSTypeAliasDeclaration", 153],
	["TSTypeAnnotation", 154],
	["TSTypeAssertion", 155],
	["TSTypeLiteral", 156],
	["TSTypeOperator", 157],
	["TSTypeParameter", 158],
	["TSTypeParameterDeclaration", 159],
	["TSTypeParameterInstantiation", 160],
	["TSTypePredicate", 161],
	["TSTypeQuery", 162],
	["TSTypeReference", 163],
	["TSUnionType", 164]
]);
//#endregion
//#region ../../node_modules/.pnpm/oxc-parser@0.139.0/node_modules/oxc-parser/src-js/visit/visitor.js
let compiledVisitor;
function createCompiledVisitor() {
	compiledVisitor = [];
	for (let i = 165; i !== 0; i--) compiledVisitor.push(null);
	return compiledVisitor;
}
const mergedLeafVisitorTypeIds = [];
const mergedEnterVisitorTypeIds = [];
const mergedExitVisitorTypeIds = [];
for (let i = 27; i !== 0; i--) mergedLeafVisitorTypeIds.push(0);
for (let i = 138; i !== 0; i--) {
	mergedEnterVisitorTypeIds.push(0);
	mergedExitVisitorTypeIds.push(0);
}
mergedLeafVisitorTypeIds.length = 0;
mergedEnterVisitorTypeIds.length = 0;
mergedExitVisitorTypeIds.length = 0;
let hasActiveVisitors = false;
const enterExitObjectCache = [];
let enterExitObjectCacheNextIndex = 0;
function getEnterExitObject() {
	if (enterExitObjectCacheNextIndex < enterExitObjectCache.length) return enterExitObjectCache[enterExitObjectCacheNextIndex++];
	const enterExit = {
		enter: null,
		exit: null
	};
	enterExitObjectCache.push(enterExit);
	enterExitObjectCacheNextIndex++;
	return enterExit;
}
const visitFnArrayCache = [];
let visitFnArrayCacheNextIndex = 0;
function createVisitFnArray(visit1, visit2) {
	if (visitFnArrayCacheNextIndex < visitFnArrayCache.length) {
		const arr = visitFnArrayCache[visitFnArrayCacheNextIndex++];
		arr.push(visit1, visit2);
		return arr;
	}
	const arr = [visit1, visit2];
	visitFnArrayCache.push(arr);
	visitFnArrayCacheNextIndex++;
	return arr;
}
/**
* Add a visitor to compiled visitor.
*
* @param visitor - Visitor object
*/
function addVisitorToCompiled(visitor) {
	if (visitor === null || typeof visitor !== "object") throw new TypeError("Visitor must be an object");
	const keys = Object.keys(visitor), keysLen = keys.length;
	if (keysLen === 0) return;
	hasActiveVisitors = true;
	for (let i = 0; i < keysLen; i++) {
		let name = keys[i];
		const visitFn = visitor[name];
		if (typeof visitFn !== "function") throw new TypeError(`'${name}' property of visitor object is not a function`);
		const isExit = name.endsWith(":exit");
		if (isExit) name = name.slice(0, -5);
		const typeId = NODE_TYPE_IDS_MAP.get(name);
		if (typeId === void 0) throw new Error(`Unknown node type '${name}' in visitor object`);
		const existing = compiledVisitor[typeId];
		if (typeId < 27) if (existing === null) compiledVisitor[typeId] = visitFn;
		else if (Array.isArray(existing)) if (isExit) existing.push(visitFn);
		else existing.splice(existing.length - 1, 0, visitFn);
		else {
			compiledVisitor[typeId] = isExit ? createVisitFnArray(existing, visitFn) : createVisitFnArray(visitFn, existing);
			mergedLeafVisitorTypeIds.push(typeId);
		}
		else if (existing === null) {
			const enterExit = compiledVisitor[typeId] = getEnterExitObject();
			if (isExit) enterExit.exit = visitFn;
			else enterExit.enter = visitFn;
		} else if (isExit) {
			const { exit } = existing;
			if (exit === null) existing.exit = visitFn;
			else if (Array.isArray(exit)) exit.push(visitFn);
			else {
				existing.exit = createVisitFnArray(exit, visitFn);
				mergedExitVisitorTypeIds.push(typeId);
			}
		} else {
			const { enter } = existing;
			if (enter === null) existing.enter = visitFn;
			else if (Array.isArray(enter)) enter.push(visitFn);
			else {
				existing.enter = createVisitFnArray(enter, visitFn);
				mergedEnterVisitorTypeIds.push(typeId);
			}
		}
	}
}
/**
* Finalize compiled visitor.
*
* After calling this function, `compiledVisitor` is ready to be used to walk the AST.
*
* @returns {boolean} - `true` if compiled visitor visits at least 1 AST type
*/
function finalizeCompiledVisitor() {
	if (hasActiveVisitors === false) return false;
	for (let i = mergedLeafVisitorTypeIds.length - 1; i >= 0; i--) {
		const typeId = mergedLeafVisitorTypeIds[i];
		compiledVisitor[typeId] = mergeVisitFns(compiledVisitor[typeId]);
	}
	for (let i = mergedEnterVisitorTypeIds.length - 1; i >= 0; i--) {
		const typeId = mergedEnterVisitorTypeIds[i];
		const enterExit = compiledVisitor[typeId];
		enterExit.enter = mergeVisitFns(enterExit.enter);
	}
	for (let i = mergedExitVisitorTypeIds.length - 1; i >= 0; i--) {
		const typeId = mergedExitVisitorTypeIds[i];
		const enterExit = compiledVisitor[typeId];
		enterExit.exit = mergeVisitFns(enterExit.exit);
	}
	mergedLeafVisitorTypeIds.length = 0;
	mergedEnterVisitorTypeIds.length = 0;
	mergedExitVisitorTypeIds.length = 0;
	visitFnArrayCacheNextIndex = 0;
	hasActiveVisitors = false;
	return true;
}
/**
* Merge array of visit functions into a single function, which calls each of input functions in turn.
*
* The array passed is cleared (length set to 0), so the array can be reused.
*
* The merged function is statically defined and does not contain a loop, to hopefully allow
* JS engine to heavily optimize it.
*
* `mergers` contains pre-defined functions to merge up to 5 visit functions.
* Merger functions for merging more than 5 visit functions are created dynamically on demand.
*
* @param visitFns - Array of visit functions
* @returns Function which calls all of `visitFns` in turn.
*/
function mergeVisitFns(visitFns) {
	const numVisitFns = visitFns.length;
	let merger;
	if (mergers.length <= numVisitFns) {
		while (mergers.length < numVisitFns) mergers.push(null);
		merger = createMerger(numVisitFns);
		mergers.push(merger);
	} else {
		merger = mergers[numVisitFns];
		if (merger === null) merger = mergers[numVisitFns] = createMerger(numVisitFns);
	}
	const mergedFn = merger(...visitFns);
	visitFns.length = 0;
	return mergedFn;
}
/**
* Create a merger function that merges `fnCount` functions.
*
* @param fnCount - Number of functions to be merged
* @returns Function to merge `fnCount` functions
*/
function createMerger(fnCount) {
	const args = [];
	let body = "return node=>{";
	for (let i = 1; i <= fnCount; i++) {
		args.push(`visit${i}`);
		body += `visit${i}(node);`;
	}
	body += "}";
	args.push(body);
	return new Function(...args);
}
const mergers = [
	null,
	null,
	(visit1, visit2) => (node) => {
		visit1(node);
		visit2(node);
	},
	(visit1, visit2, visit3) => (node) => {
		visit1(node);
		visit2(node);
		visit3(node);
	},
	(visit1, visit2, visit3, visit4) => (node) => {
		visit1(node);
		visit2(node);
		visit3(node);
		visit4(node);
	},
	(visit1, visit2, visit3, visit4, visit5) => (node) => {
		visit1(node);
		visit2(node);
		visit3(node);
		visit4(node);
		visit5(node);
	}
];
//#endregion
//#region src/utils/visitor.ts
/**
* Visitor class for traversing AST.
*
* @example
* ```ts
* import { Visitor } from 'rolldown/utils';
* import { parseSync } from 'rolldown/utils';
*
* const result = parseSync(...);
* const visitor = new Visitor({
*   VariableDeclaration(path) {
*     // Do something with the variable declaration
*   },
*   "VariableDeclaration:exit"(path) {
*     // Do something after visiting the variable declaration
*   }
* });
* visitor.visit(result.program);
* ```
*
* @category Utilities
* @experimental
*/
var Visitor = class {
	#compiledVisitor = null;
	constructor(visitor) {
		const compiledVisitor = createCompiledVisitor();
		addVisitorToCompiled(visitor);
		if (finalizeCompiledVisitor()) this.#compiledVisitor = compiledVisitor;
	}
	visit(program) {
		const compiledVisitor = this.#compiledVisitor;
		if (compiledVisitor !== null) walkProgram(program, compiledVisitor);
	}
};
//#endregion
export { TsconfigCache, Visitor, minify, minifySync, parse, parseSync, transform, transformSync };
e);
}
function walkTSInstantiationExpression(node, visitors) {
	let enterExit = visitors[129], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.expression, visitors);
	walkNode(node.typeArguments, visitors);
	exit !== null && exit(node);
}
function walkTSInterfaceBody(node, visitors) {
	let enterExit = visitors[130], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.body, visitors);
	exit !== null && exit(node);
}
function walkTSInterfaceDeclaration(node, visitors) {
	let enterExit = visitors[131], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.id, visitors);
	walkNode(node.typeParameters, visitors);
	walkNode(node.extends, visitors);
	walkNode(node.body, visitors);
	exit !== null && exit(node);
}
function walkTSInterfaceHeritage(node, visitors) {
	let enterExit = visitors[132], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.expression, visitors);
	walkNode(node.typeArguments, visitors);
	exit !== null && exit(node);
}
function walkTSIntersectionType(node, visitors) {
	let enterExit = visitors[133], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.types, visitors);
	exit !== null && exit(node);
}
function walkTSJSDocNonNullableType(node, visitors) {
	let enterExit = visitors[134], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.typeAnnotation, visitors);
	exit !== null && exit(node);
}
function walkTSJSDocNullableType(node, visitors) {
	let enterExit = visitors[135], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.typeAnnotation, visitors);
	exit !== null && exit(node);
}
function walkTSLiteralType(node, visitors) {
	let enterExit = visitors[136], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.literal, visitors);
	exit !== null && exit(node);
}
function walkTSMappedType(node, visitors) {
	let enterExit = visitors[137], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.key, visitors);
	walkNode(node.constraint, visitors);
	walkNode(node.nameType, visitors);
	walkNode(node.typeAnnotation, visitors);
	exit !== null && exit(node);
}
function walkTSMethodSignature(node, visitors) {
	let enterExit = visitors[138], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.key, visitors);
	walkNode(node.typeParameters, visitors);
	walkNode(node.params, visitors);
	walkNode(node.returnType, visitors);
	exit !== null && exit(node);
}
function walkTSModuleBlock(node, visitors) {
	let enterExit = visitors[139], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.body, visitors);
	exit !== null && exit(node);
}
function walkTSModuleDeclaration(node, visitors) {
	let enterExit = visitors[140], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.id, visitors);
	walkNode(node.body, visitors);
	exit !== null && exit(node);
}
function walkTSNamedTupleMember(node, visitors) {
	let enterExit = visitors[141], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.label, visitors);
	walkNode(node.elementType, visitors);
	exit !== null && exit(node);
}
function walkTSNamespaceExportDeclaration(node, visitors) {
	let enterExit = visitors[142], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.id, visitors);
	exit !== null && exit(node);
}
function walkTSNonNullExpression(node, visitors) {
	let enterExit = visitors[143], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.expression, visitors);
	exit !== null && exit(node);
}
function walkTSOptionalType(node, visitors) {
	let enterExit = visitors[144], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.typeAnnotation, visitors);
	exit !== null && exit(node);
}
function walkTSParameterProperty(node, visitors) {
	let enterExit = visitors[145], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.decorators, visitors);
	walkNode(node.parameter, visitors);
	exit !== null && exit(node);
}
function walkTSParenthesizedType(node, visitors) {
	let enterExit = visitors[146], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.typeAnnotation, visitors);
	exit !== null && exit(node);
}
function walkTSPropertySignature(node, visitors) {
	let enterExit = visitors[147], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.key, visitors);
	walkNode(node.typeAnnotation, visitors);
	exit !== null && exit(node);
}
function walkTSQualifiedName(node, visitors) {
	let enterExit = visitors[148], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.left, visitors);
	walkNode(node.right, visitors);
	exit !== null && exit(node);
}
function walkTSRestType(node, visitors) {
	let enterExit = visitors[149], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.typeAnnotation, visitors);
	exit !== null && exit(node);
}
function walkTSSatisfiesExpression(node, visitors) {
	let enterExit = visitors[150], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.expression, visitors);
	walkNode(node.typeAnnotation, visitors);
	exit !== null && exit(node);
}
function walkTSTemplateLiteralType(node, visitors) {
	let enterExit = visitors[151], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.quasis, visitors);
	walkNode(node.types, visitors);
	exit !== null && exit(node);
}
function walkTSTupleType(node, visitors) {
	let enterExit = visitors[152], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.elementTypes, visitors);
	exit !== null && exit(node);
}
function walkTSTypeAliasDeclaration(node, visitors) {
	let enterExit = visitors[153], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.id, visitors);
	walkNode(node.typeParameters, visitors);
	walkNode(node.typeAnnotation, visitors);
	exit !== null && exit(node);
}
function walkTSTypeAnnotation(node, visitors) {
	let enterExit = visitors[154], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.typeAnnotation, visitors);
	exit !== null && exit(node);
}
function walkTSTypeAssertion(node, visitors) {
	let enterExit = visitors[155], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.typeAnnotation, visitors);
	walkNode(node.expression, visitors);
	exit !== null && exit(node);
}
function walkTSTypeLiteral(node, visitors) {
	let enterExit = visitors[156], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.members, visitors);
	exit !== null && exit(node);
}
function walkTSTypeOperator(node, visitors) {
	let enterExit = visitors[157], exit = null, enter;
	if (enterExit !== null) {
		({enter, exit} = enterExit);
		enter !== null && enter(node);
	}
	walkNode(node.typeAnnotation, visitors);
	exit !== null && exit(node);
}
function walkTSTypeParameter(node, visitors) {
	let enterExit = visi
