// rolldown/dist/shared/prompt--dNycKSZ.mjs
is.render.bind(this), this._render = i.bind(this), this._track = F, this._abortSignal = D, this.input = e, this.output = s;
	}
	unsubscribe() {
		this._subscribers.clear();
	}
	setSubscriber(u, F) {
		const e = this._subscribers.get(u) ?? [];
		e.push(F), this._subscribers.set(u, e);
	}
	on(u, F) {
		this.setSubscriber(u, { cb: F });
	}
	once(u, F) {
		this.setSubscriber(u, {
			cb: F,
			once: true
		});
	}
	emit(u, ...F) {
		const e = this._subscribers.get(u) ?? [], s = [];
		for (const i of e) i.cb(...F), i.once && s.push(() => e.splice(e.indexOf(i), 1));
		for (const i of s) i();
	}
	prompt() {
		return new Promise((u, F) => {
			if (this._abortSignal) {
				if (this._abortSignal.aborted) return this.state = "cancel", this.close(), u(S);
				this._abortSignal.addEventListener("abort", () => {
					this.state = "cancel", this.close();
				}, { once: true });
			}
			const e = new WriteStream(0);
			e._write = (s, i, D) => {
				this._track && (this.value = this.rl?.line.replace(/\t/g, ""), this._cursor = this.rl?.cursor ?? 0, this.emit("value", this.value)), D();
			}, this.input.pipe(e), this.rl = f.createInterface({
				input: this.input,
				output: e,
				tabSize: 2,
				prompt: "",
				escapeCodeTimeout: 50
			}), f.emitKeypressEvents(this.input, this.rl), this.rl.prompt(), this.opts.initialValue !== void 0 && this._track && this.rl.write(this.opts.initialValue), this.input.on("keypress", this.onKeypress), d$1(this.input, true), this.output.on("resize", this.render), this.render(), this.once("submit", () => {
				this.output.write(srcExports.cursor.show), this.output.off("resize", this.render), d$1(this.input, false), u(this.value);
			}), this.once("cancel", () => {
				this.output.write(srcExports.cursor.show), this.output.off("resize", this.render), d$1(this.input, false), u(S);
			});
		});
	}
	onKeypress(u, F) {
		if (this.state === "error" && (this.state = "active"), F?.name && (!this._track && c.aliases.has(F.name) && this.emit("cursor", c.aliases.get(F.name)), c.actions.has(F.name) && this.emit("cursor", F.name)), u && (u.toLowerCase() === "y" || u.toLowerCase() === "n") && this.emit("confirm", u.toLowerCase() === "y"), u === "	" && this.opts.placeholder && (this.value || (this.rl?.write(this.opts.placeholder), this.emit("value", this.opts.placeholder))), u && this.emit("key", u.toLowerCase()), F?.name === "return") {
			if (this.opts.validate) {
				const e = this.opts.validate(this.value);
				e && (this.error = e instanceof Error ? e.message : e, this.state = "error", this.rl?.write(this.value));
			}
			this.state !== "error" && (this.state = "submit");
		}
		k$1([
			u,
			F?.name,
			F?.sequence
		], "cancel") && (this.state = "cancel"), (this.state === "submit" || this.state === "cancel") && this.emit("finalize"), this.render(), (this.state === "submit" || this.state === "cancel") && this.close();
	}
	close() {
		this.input.unpipe(), this.input.removeListener("keypress", this.onKeypress), this.output.write(`
`), d$1(this.input, false), this.rl?.close(), this.rl = void 0, this.emit(`${this.state}`, this.value), this.unsubscribe();
	}
	restoreCursor() {
		const u = G(this._prevFrame, process.stdout.columns, { hard: true }).split(`
`).length - 1;
		this.output.write(srcExports.cursor.move(-999, u * -1));
	}
	render() {
		const u = G(this._render(this) ?? "", process.stdout.columns, { hard: true });
		if (u !== this._prevFrame) {
			if (this.state === "initial") this.output.write(srcExports.cursor.hide);
			else {
				const F = lD(this._prevFrame, u);
				if (this.restoreCursor(), F && F?.length === 1) {
					const e = F[0];
					this.output.write(srcExports.cursor.move(0, e)), this.output.write(srcExports.erase.lines(1));
					const s = u.split(`
`);
					this.output.write(s[e]), this._prevFrame = u, this.output.write(srcExports.cursor.move(0, s.length - e - 1));
					return;
				}
				if (F && F?.length > 1) {
					const e = F[0];
					this.output.write(srcExports.cursor.move(0, e)), this.output.write(srcExports.erase.down());
					const s = u.split(`
`).slice(e);
					this.output.write(s.join(`
`)), this._prevFrame = u;
					return;
				}
				this.output.write(srcExports.erase.down());
			}
			this.output.write(u), this.state === "initial" && (this.state = "active"), this._prevFrame = u;
		}
	}
};
var fD = class extends x {
	get cursor() {
		return this.value ? 0 : 1;
	}
	get _value() {
		return this.cursor === 0;
	}
	constructor(u) {
		super(u, false), this.value = !!u.initialValue, this.on("value", () => {
			this.value = this._value;
		}), this.on("confirm", (F) => {
			this.output.write(srcExports.cursor.move(0, -1)), this.value = F, this.state = "submit", this.close();
		}), this.on("cursor", () => {
			this.value = !this.value;
		});
	}
};
var bD = Object.defineProperty;
var mD = (t, u, F) => u in t ? bD(t, u, {
	enumerable: true,
	configurable: true,
	writable: true,
	value: F
}) : t[u] = F;
var Y = (t, u, F) => (mD(t, typeof u != "symbol" ? u + "" : u, F), F);
let wD = class extends x {
	constructor(u) {
		super(u, false), Y(this, "options"), Y(this, "cursor", 0), this.options = u.options, this.value = [...u.initialValues ?? []], this.cursor = Math.max(this.options.findIndex(({ value: F }) => F === u.cursorAt), 0), this.on("key", (F) => {
			F === "a" && this.toggleAll();
		}), this.on("cursor", (F) => {
			switch (F) {
				case "left":
				case "up":
					this.cursor = this.cursor === 0 ? this.options.length - 1 : this.cursor - 1;
					break;
				case "down":
				case "right":
					this.cursor = this.cursor === this.options.length - 1 ? 0 : this.cursor + 1;
					break;
				case "space":
					this.toggleValue();
					break;
			}
		});
	}
	get _value() {
		return this.options[this.cursor].value;
	}
	toggleAll() {
		const u = this.value.length === this.options.length;
		this.value = u ? [] : this.options.map((F) => F.value);
	}
	toggleValue() {
		const u = this.value.includes(this._value);
		this.value = u ? this.value.filter((F) => F !== this._value) : [...this.value, this._value];
	}
};
var SD = Object.defineProperty;
var $D = (t, u, F) => u in t ? SD(t, u, {
	enumerable: true,
	configurable: true,
	writable: true,
	value: F
}) : t[u] = F;
var q = (t, u, F) => ($D(t, typeof u != "symbol" ? u + "" : u, F), F);
var jD = class extends x {
	constructor(u) {
		super(u, false), q(this, "options"), q(this, "cursor", 0), this.options = u.options, this.cursor = this.options.findIndex(({ value: F }) => F === u.initialValue), this.cursor === -1 && (this.cursor = 0), this.changeValue(), this.on("cursor", (F) => {
			switch (F) {
				case "left":
				case "up":
					this.cursor = this.cursor === 0 ? this.options.length - 1 : this.cursor - 1;
					break;
				case "down":
				case "right":
					this.cursor = this.cursor === this.options.length - 1 ? 0 : this.cursor + 1;
					break;
			}
			this.changeValue();
		});
	}
	get _value() {
		return this.options[this.cursor];
	}
	changeValue() {
		this.value = this._value.value;
	}
};
var PD = class extends x {
	get valueWithCursor() {
		if (this.state === "submit") return this.value;
		if (this.cursor >= this.value.length) return `${this.value}\u2588`;
		const u = this.value.slice(0, this.cursor), [F, ...e$1] = this.value.slice(this.cursor);
		return `${u}${e.inverse(F)}${e$1.join("")}`;
	}
	get cursor() {
		return this._cursor;
	}
	constructor(u) {
		super(u), this.on("finalize", () => {
			this.value || (this.value = u.defaultValue);
		});
	}
};
function ce() {
	return g$1.platform !== "win32" ? g$1.env.TERM !== "linux" : !!g$1.env.CI || !!g$1.env.WT_SESSION || !!g$1.env.TERMINUS_SUBLIME || g$1.env.ConEmuTask === "{cmd::Cmder}" || g$1.env.TERM_PROGRAM === "Terminus-Sublime" || g$1.env.TERM_PROGRAM === "vscode" || g$1.env.TERM === "xterm-256color" || g$1.env.TERM === "alacritty" || g$1.env.TERMINAL_EMULATOR === "JetBrains-JediTerm";
}
const V = ce();
const u = (t, n) => V ? t : n;
const le = u("❯", ">");
const L = u("■", "x");
const W = u("▲", "x");
const C = u("✔", "√");
const o = u("");
const d = u("");
const k = u("●", ">");
const P = u("○", " ");
const A = u("◻", "[•]");
const T = u("◼", "[+]");
const F = u("◻", "[ ]");
const w = (t) => {
	switch (t) {
		case "initial":
		case "active": return e.cyan(le);
		case "cancel": return e.red(L);
		case "error": return e.yellow(W);
		case "submit": return e.green(C);
	}
};
const B = (t) => {
	const { cursor: n, options: s, style: r } = t, i = t.maxItems ?? Number.POSITIVE_INFINITY, a = Math.max(process.stdout.rows - 4, 0), c = Math.min(a, Math.max(i, 5));
	let l = 0;
	n >= l + c - 3 ? l = Math.max(Math.min(n - c + 3, s.length - c), 0) : n < l + 2 && (l = Math.max(n - 2, 0));
	const $ = c < s.length && l > 0, p = c < s.length && l + c < s.length;
	return s.slice(l, l + c).map((M, v, x) => {
		const j = v === 0 && $, E = v === x.length - 1 && p;
		return j || E ? e.dim("...") : r(M, v + l === n);
	});
};
const he = (t) => new PD({
	validate: t.validate,
	placeholder: t.placeholder,
	defaultValue: t.defaultValue,
	initialValue: t.initialValue,
	render() {
		const n = `${e.gray(o)}
${w(this.state)} ${t.message}
`, s = t.placeholder ? e.inverse(t.placeholder[0]) + e.dim(t.placeholder.slice(1)) : e.inverse(e.hidden("_")), r = this.value ? this.valueWithCursor : s;
		switch (this.state) {
			case "error": return `${n.trim()}
${e.yellow(o)} ${r}
${e.yellow(d)} ${e.yellow(this.error)}
`;
			case "submit": return `${n}${e.gray(o)} ${e.dim(this.value || t.placeholder)}`;
			case "cancel": return `${n}${e.gray(o)} ${e.strikethrough(e.dim(this.value ?? ""))}${this.value?.trim() ? `
${e.gray(o)}` : ""}`;
			default: return `${n}${e.cyan(o)} ${r}
${e.cyan(d)}
`;
		}
	}
}).prompt();
const ye = (t) => {
	const n = t.active ?? "Yes", s = t.inactive ?? "No";
	return new fD({
		active: n,
		inactive: s,
		initialValue: t.initialValue ?? true,
		render() {
			const r = `${e.gray(o)}
${w(this.state)} ${t.message}
`, i = this.value ? n : s;
			switch (this.state) {
				case "submit": return `${r}${e.gray(o)} ${e.dim(i)}`;
				case "cancel": return `${r}${e.gray(o)} ${e.strikethrough(e.dim(i))}
${e.gray(o)}`;
				default: return `${r}${e.cyan(o)} ${this.value ? `${e.green(k)} ${n}` : `${e.dim(P)} ${e.dim(n)}`} ${e.dim("/")} ${this.value ? `${e.dim(P)} ${e.dim(s)}` : `${e.green(k)} ${s}`}
${e.cyan(d)}
`;
			}
		}
	}).prompt();
};
const ve = (t) => {
	const n = (s, r) => {
		const i = s.label ?? String(s.value);
		switch (r) {
			case "selected": return `${e.dim(i)}`;
			case "active": return `${e.green(k)} ${i} ${s.hint ? e.dim(`(${s.hint})`) : ""}`;
			case "cancelled": return `${e.strikethrough(e.dim(i))}`;
			default: return `${e.dim(P)} ${e.dim(i)}`;
		}
	};
	return new jD({
		options: t.options,
		initialValue: t.initialValue,
		render() {
			const s = `${e.gray(o)}
${w(this.state)} ${t.message}
`;
			switch (this.state) {
				case "submit": return `${s}${e.gray(o)} ${n(this.options[this.cursor], "selected")}`;
				case "cancel": return `${s}${e.gray(o)} ${n(this.options[this.cursor], "cancelled")}
${e.gray(o)}`;
				default: return `${s}${e.cyan(o)} ${B({
					cursor: this.cursor,
					options: this.options,
					maxItems: t.maxItems,
					style: (r, i) => n(r, i ? "active" : "inactive")
				}).join(`
${e.cyan(o)}  `)}
${e.cyan(d)}
`;
			}
		}
	}).prompt();
};
const fe = (t) => {
	const n = (s, r) => {
		const i = s.label ?? String(s.value);
		return r === "active" ? `${e.cyan(A)} ${i} ${s.hint ? e.dim(`(${s.hint})`) : ""}` : r === "selected" ? `${e.green(T)} ${e.dim(i)}` : r === "cancelled" ? `${e.strikethrough(e.dim(i))}` : r === "active-selected" ? `${e.green(T)} ${i} ${s.hint ? e.dim(`(${s.hint})`) : ""}` : r === "submitted" ? `${e.dim(i)}` : `${e.dim(F)} ${e.dim(i)}`;
	};
	return new wD({
		options: t.options,
		initialValues: t.initialValues,
		required: t.required ?? true,
		cursorAt: t.cursorAt,
		validate(s) {
			if (this.required && s.length === 0) return `Please select at least one option.
${e.reset(e.dim(`Press ${e.gray(e.bgWhite(e.inverse(" space ")))} to select, ${e.gray(e.bgWhite(e.inverse(" enter ")))} to submit`))}`;
		},
		render() {
			const s = `${e.gray(o)}
${w(this.state)} ${t.message}
`, r = (i, a) => {
				const c = this.value.includes(i.value);
				return a && c ? n(i, "active-selected") : c ? n(i, "selected") : n(i, a ? "active" : "inactive");
			};
			switch (this.state) {
				case "submit": return `${s}${e.gray(o)} ${this.options.filter(({ value: i }) => this.value.includes(i)).map((i) => n(i, "submitted")).join(e.dim(", ")) || e.dim("none")}`;
				case "cancel": {
					const i = this.options.filter(({ value: a }) => this.value.includes(a)).map((a) => n(a, "cancelled")).join(e.dim(", "));
					return `${s}${e.gray(o)} ${i.trim() ? `${i}
${e.gray(o)}` : ""}`;
				}
				case "error": {
					const i = this.error.split(`
`).map((a, c) => c === 0 ? `${e.yellow(d)} ${e.yellow(a)}` : `   ${a}`).join(`
`);
					return `${s + e.yellow(o)} ${B({
						options: this.options,
						cursor: this.cursor,
						maxItems: t.maxItems,
						style: r
					}).join(`
${e.yellow(o)}  `)}
${i}
`;
				}
				default: return `${s}${e.cyan(o)} ${B({
					options: this.options,
					cursor: this.cursor,
					maxItems: t.maxItems,
					style: r
				}).join(`
${e.cyan(o)}  `)}
${e.cyan(d)}
`;
			}
		}
	}).prompt();
};
`${e.gray(o)}`;
const kCancel = Symbol.for("cancel");
async function prompt(message, opts = {}) {
	const handleCancel = (value) => {
		if (typeof value !== "symbol" || value.toString() !== "Symbol(clack:cancel)") return value;
		switch (opts.cancel) {
			case "reject": {
				const error = /* @__PURE__ */ new Error("Prompt cancelled.");
				error.name = "ConsolaPromptCancelledError";
				if (Error.captureStackTrace) Error.captureStackTrace(error, prompt);
				throw error;
			}
			case "undefined": return;
			case "null": return null;
			case "symbol": return kCancel;
			default:
			case "default": return opts.default ?? opts.initial;
		}
	};
	if (!opts.type || opts.type === "text") return await he({
		message,
		defaultValue: opts.default,
		placeholder: opts.placeholder,
		initialValue: opts.initial
	}).then(handleCancel);
	if (opts.type === "confirm") return await ye({
		message,
		initialValue: opts.initial
	}).then(handleCancel);
	if (opts.type === "select") return await ve({
		message,
		options: opts.options.map((o) => typeof o === "string" ? {
			value: o,
			label: o
		} : o),
		initialValue: opts.initial
	}).then(handleCancel);
	if (opts.type === "multiselect") return await fe({
		message,
		options: opts.options.map((o) => typeof o === "string" ? {
			value: o,
			label: o
		} : o),
		required: opts.required,
		initialValues: opts.initial
	}).then(handleCancel);
	throw new Error(`Unknown prompt type: ${opts.type}`);
}
//#endregion
export { prompt };
e(/\r\n/g, `
`).split(`
`).map((e) => oD(e, u, F)).join(`
`);
}
const c = {
	actions: /* @__PURE__ */ new Set([
		"up",
		"down",
		"left",
		"right",
		"space",
		"enter",
		"cancel"
	]),
	aliases: /* @__PURE__ */ new Map([
		["k", "up"],
		["j", "down"],
		["h", "left"],
		["l", "right"],
		["", "cancel"],
		["escape", "cancel"]
	])
};
function k$1(t, u) {
	if (typeof t == "string") return c.aliases.get(t) === u;
	for (const F of t) if (F !== void 0 && k$1(F, u)) return true;
	return false;
}
function lD(t, u) {
	if (t === u) return;
	const F = t.split(`
`), e = u.split(`
`), s = [];
	for (let i = 0; i < Math.max(F.length, e.length); i++) F[i] !== e[i] && s.push(i);
	return s;
}
globalThis.process.platform.startsWith("win");
const S = Symbol("clack:cancel");
function d$1(t, u) {
	const F = t;
	F.isTTY && F.setRawMode(u);
}
var AD = Object.defineProperty;
var pD = (t, u, F) => u in t ? AD(t, u, {
	enumerable: true,
	configurable: true,
	writable: true,
	value: F
}) : t[u] = F;
var h = (t, u, F) => (pD(t, typeof u != "symbol" ? u + "" : u, F), F);
var x = class {
	constructor(u, F = true) {
		h(this, "input"), h(this, "output"), h(this, "_abortSignal"), h(this, "rl"), h(this, "opts"), h(this, "_render"), h(this, "_track", false), h(this, "_prevFrame", ""), h(this, "_subscribers", /* @__PURE__ */ new Map()), h(this, "_cursor", 0), h(this, "state", "initial"), h(this, "error", ""), h(this, "value");
		const { input: e = stdin, output: s = stdout, render: i, signal: D, ...C } = u;
		this.opts = C, this.onKeypress = this.onKeypress.bind(this), this.close = this.close.bind(this), this.render = th
